File: sdk-java/src/test/java/io/littlehorse/sdk/wfsdk/internal/taskdefutil/LHTaskParameterTest.java
package io.littlehorse.sdk.wfsdk.internal.taskdefutil;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import io.littlehorse.sdk.common.adapter.LHStringAdapter;
import io.littlehorse.sdk.common.adapter.LHTypeAdapterRegistry;
import io.littlehorse.sdk.common.exception.TaskSchemaMismatchError;
import io.littlehorse.sdk.common.proto.InlineArrayDef;
import io.littlehorse.sdk.common.proto.InlineStruct;
import io.littlehorse.sdk.common.proto.StructDefId;
import io.littlehorse.sdk.common.proto.TypeDefinition;
import io.littlehorse.sdk.common.proto.VariableDef;
import io.littlehorse.sdk.common.proto.VariableType;
import io.littlehorse.sdk.testutils.TestReflection;
import io.littlehorse.sdk.worker.LHTaskMethod;
import io.littlehorse.sdk.worker.LHType;
import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.util.Map;
import java.util.UUID;
import org.junit.jupiter.api.Test;

public class LHTaskParameterTest {
    public class ParameterTestTasks {
        @LHTaskMethod("test-primitive")
        public void primitiveParamTask(@LHType(name = "param1") String param1) {}

        @LHTaskMethod("test-json-arr")
        public void jsonParamTask(@LHType(name = "param1") String[] param1) {}

        @LHTaskMethod("test-native-arr")
        public void nativeArrayParamTask(@LHType(name = "param1", isLHArray = true) String[] param1) {}

        @LHTaskMethod("test-type-adapted-method")
        public void typeAdaptedParamTask(@LHType(name = "param1") UUID param1) {}

        @LHTaskMethod("test-masked-param")
        public void maskedParamTask(@LHType(name = "param1", masked = true) int param1) {}

        @LHTaskMethod("test-inline-struct-param")
        public void inlineStructTask(@LHType(name = "param1", structDefName = "customer") InlineStruct param1) {}

        @LHTaskMethod("inline-struct-placeholder-task")
        public void inlineStructPlaceholderTask(
                @LHType(name = "param1", structDefName = "${inputStruct}") InlineStruct customer) {}

        @LHTaskMethod("inline-struct-invalid-task")
        public void inlineStructInvalidTask(InlineStruct customer) {}
        ;

        @LHTaskMethod("non-inline-structdef-invalid-task")
        public void nonInlineStructDefInvalidTask(@LHType(structDefName = "customer") String customer) {}

        @LHTaskMethod("non-array-lh-array-invalid-task")
        public void nonArrayLhArrayInvalidTask(@LHType(isLHArray = true) String customer) {}
    }

    @Test
    public void shouldHandlePrimitiveTaskParameter() {
        Method taskMethod = TestReflection.getTaskMethodByName(ParameterTestTasks.class, "test-primitive");
        Parameter parameter = taskMethod.getParameters()[0];
        LHTaskParameter taskParameter = new LHTaskParameter(parameter, LHTypeAdapterRegistry.empty(), Map.of());

        VariableDef actualVariableDef = taskParameter.getVariableDef();
        VariableDef expectedVariableDef = VariableDef.newBuilder()
                .setName("param1")
                .setTypeDef(TypeDefinition.newBuilder()
                        .setPrimitiveType(VariableType.STR)
                        .build())
                .build();

        assertThat(actualVariableDef).isEqualTo(expectedVariableDef);
    }

    @Test
    public void shouldHandleJsonArrayTaskParameter() {
        Method taskMethod = TestReflection.getTaskMethodByName(ParameterTestTasks.class, "test-json-arr");
        Parameter parameter = taskMethod.getParameters()[0];
        LHTaskParameter taskParameter = new LHTaskParameter(parameter, LHTypeAdapterRegistry.empty(), Map.of());

        VariableDef actualVariableDef = taskParameter.getVariableDef();
        VariableDef expectedVariableDef = VariableDef.newBuilder()
                .setName("param1")
                .setTypeDef(TypeDefinition.newBuilder()
                        .setPrimitiveType(VariableType.JSON_ARR)
                        .build())
                .build();

        assertThat(actualVariableDef).isEqualTo(expectedVariableDef);
    }

    @Test
    public void shouldHandleNativeArrayTaskParameter() {
        Method taskMethod = TestReflection.getTaskMethodByName(ParameterTestTasks.class, "test-native-arr");
        Parameter parameter = taskMethod.getParameters()[0];
        LHTaskParameter taskParameter = new LHTaskParameter(parameter, LHTypeAdapterRegistry.empty(), Map.of());

        VariableDef actualVariableDef = taskParameter.getVariableDef();
        VariableDef expectedVariableDef = VariableDef.newBuilder()
                .setName("param1")
                .setTypeDef(TypeDefinition.newBuilder()
                        .setInlineArrayDef(InlineArrayDef.newBuilder()
                                .setArrayType(TypeDefinition.newBuilder()
                                        .setPrimitiveType(VariableType.STR)
                                        .build()))
                        .build())
                .build();

        assertThat(actualVariableDef).isEqualTo(expectedVariableDef);
    }

    @Test
    public void shouldHandleTypeAdaptedLHParameterType() {
        LHTypeAdapterRegistry typeAdapterRegistry =
                LHTypeAdapterRegistry.from(Map.of(UUID.class, new LHStringAdapter<UUID>() {
                    @Override
                    public Class<UUID> getTypeClass() {
                        return UUID.class;
                    }

                    @Override
                    public String toString(UUID src) {
                        // TODO Auto-generated method stub
                        throw new UnsupportedOperationException("Unimplemented method 'toString'");
                    }

                    @Override
                    public UUID fromString(String src) {
                        // TODO Auto-generated method stub
                        throw new UnsupportedOperationException("Unimplemented method 'fromString'");
                    }
                }));

        Method taskMethod = TestReflection.getTaskMethodByName(ParameterTestTasks.class, "test-type-adapted-method");
        Parameter parameter = taskMethod.getParameters()[0];
        LHTaskParameter taskParameter = new LHTaskParameter(parameter, typeAdapterRegistry, Map.of());

        VariableDef actualVariableDef = taskParameter.getVariableDef();
        VariableDef expectedVariableDef = VariableDef.newBuilder()
                .setName("param1")
                .setTypeDef(TypeDefinition.newBuilder()
                        .setPrimitiveType(VariableType.STR)
                        .build())
                .build();

        assertThat(actualVariableDef).isEqualTo(expectedVariableDef);
    }

    @Test
    public void shouldHandleMaskedParameter() {
        Method taskMethod = TestReflection.getTaskMethodByName(ParameterTestTasks.class, "test-masked-param");
        Parameter parameter = taskMethod.getParameters()[0];
        LHTaskParameter taskParameter = new LHTaskParameter(parameter, LHTypeAdapterRegistry.empty(), Map.of());

        assertThat(taskParameter.getVariableDef().getTypeDef().getMasked()).isTrue();
    }

    @Test
    public void shouldInferInlineStructParameter() {
        Method taskMethod = TestReflection.getTaskMethodByName(ParameterTestTasks.class, "test-inline-struct-param");
        Parameter parameter = taskMethod.getParameters()[0];
        LHTaskParameter taskParameter = new LHTaskParameter(parameter, LHTypeAdapterRegistry.empty(), Map.of());

        VariableDef actualVariableDef = taskParameter.getVariableDef();
        VariableDef expectedVariableDef = VariableDef.newBuilder()
                .setName("param1")
                .setTypeDef(TypeDefinition.newBuilder()
                        .setStructDefId(StructDefId.newBuilder().setName("customer").setVersion(-1))
                        .build())
                .build();

        assertThat(actualVariableDef).isEqualTo(expectedVariableDef);
    }

    @Test
    void shouldResolvePlaceholdersForInlineStructTypes() {
        Map<String, String> placeholderValues = Map.of("inputStruct", "customer-request");

        Method taskMethod =
                TestReflection.getTaskMethodByName(ParameterTestTasks.class, "inline-struct-placeholder-task");
        Parameter parameter = taskMethod.getParameters()[0];
        LHTaskParameter taskParameter =
                new LHTaskParameter(parameter, LHTypeAdapterRegistry.empty(), placeholderValues);

        TypeDefinition inputTypeDef = taskParameter.getVariableDef().getTypeDef();

        assertThat(inputTypeDef.getStructDefId().getName()).isEqualTo("customer-request");
    }

    @Test
    void shouldFailWhenInlineStructTypeIsMissingStructDefName() {
        Method taskMethod = TestReflection.getTaskMethodByName(ParameterTestTasks.class, "inline-struct-invalid-task");
        Parameter parameter = taskMethod.getParameters()[0];

        assertThatThrownBy(() -> {
                    new LHTaskParameter(parameter, LHTypeAdapterRegistry.empty(), Map.of());
                })
                .isInstanceOf(TaskSchemaMismatchError.class)
                .hasMessageContaining("InlineStruct parameters must declare @LHType(structDefName = \"...\")");
    }

    @Test
    void shouldFailWhenStructDefNameIsUsedOnNonInlineStructParameter() {
        Method taskMethod =
                TestReflection.getTaskMethodByName(ParameterTestTasks.class, "non-inline-structdef-invalid-task");
        Parameter parameter = taskMethod.getParameters()[0];

        assertThatThrownBy(() -> {
                    new LHTaskParameter(parameter, LHTypeAdapterRegistry.empty(), Map.of());
                })
                .isInstanceOf(TaskSchemaMismatchError.class)
                .hasMessageContaining("@LHType(structDefName = ...) can only be used on InlineStruct");
    }

    @Test
    void shouldFailWhenIsLHArrayUsedOnNonArrayParameter() {
        Method taskMethod =
                TestReflection.getTaskMethodByName(ParameterTestTasks.class, "non-array-lh-array-invalid-task");
        Parameter parameter = taskMethod.getParameters()[0];

        assertThatThrownBy(() -> {
                    new LHTaskParameter(parameter, LHTypeAdapterRegistry.empty(), Map.of());
                })
                .isInstanceOf(TaskSchemaMismatchError.class)
                .hasMessageContaining("@LHType(isLHArray = true) can only be used on array parameters");
    }
}

File: sdk-dotnet/LittleHorse.Sdk.Tests/Worker/LHTaskParameterTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Threading.Tasks;
using LittleHorse.Sdk.Common.Proto;
using LittleHorse.Sdk.Exceptions;
using LittleHorse.Sdk.Worker;
using Xunit;

namespace LittleHorse.Sdk.Tests.Worker
{
    public class LHTaskParameterTest
    {
        [Fact]
        public void ShouldHandleJsonArrayTaskParameter()
        {
            MethodInfo method = GetTaskMethodByName(typeof(ParameterTestTasks), "test-json-arr");
            ParameterInfo parameter = method.GetParameters()[0];
            var taskParameter = new LHTaskParameter(parameter, 0, logger: null);

            var expectedVariableDef = new VariableDef
            {
                Name = "param1",
                TypeDef = new TypeDefinition { PrimitiveType = VariableType.JsonArr }
            };

            Assert.Equal(expectedVariableDef, taskParameter.VariableDef);
        }

        [Fact]
        public void ShouldHandleNativeArrayTaskParameter()
        {
            MethodInfo method = GetTaskMethodByName(typeof(ParameterTestTasks), "test-native-arr");
            ParameterInfo parameter = method.GetParameters()[0];
            var taskParameter = new LHTaskParameter(parameter, 0, logger: null);

            var expectedVariableDef = new VariableDef
            {
                Name = "param1",
                TypeDef = new TypeDefinition
                {
                    InlineArrayDef = new InlineArrayDef
                    {
                        ArrayType = new TypeDefinition { PrimitiveType = VariableType.Str }
                    }
                }
            };

            Assert.Equal(expectedVariableDef, taskParameter.VariableDef);
        }

        [Fact]
        public void ShouldHandleNativeListTaskParameter()
        {
            MethodInfo method = GetTaskMethodByName(typeof(ParameterTestTasks), "test-native-list");
            ParameterInfo parameter = method.GetParameters()[0];
            var taskParameter = new LHTaskParameter(parameter, 0, logger: null);

            Assert.Equal(TypeDefinition.DefinedTypeOneofCase.InlineArrayDef, taskParameter.VariableDef.TypeDef.DefinedTypeCase);
            Assert.Equal(VariableType.Str, taskParameter.VariableDef.TypeDef.InlineArrayDef.ArrayType.PrimitiveType);
        }

        [Fact]
        public void ShouldFailWhenIsLHArrayUsedOnNonArrayParameter()
        {
            MethodInfo method = GetTaskMethodByName(typeof(ParameterTestTasks), "non-array-lh-array-invalid-task");
            ParameterInfo parameter = method.GetParameters()[0];

            var ex = Assert.Throws<LHTaskSchemaMismatchException>(() => new LHTaskParameter(parameter, 0, logger: null));
            Assert.Contains("@LHType(isLHArray = true) can only be used on array or IList<T> parameters", ex.Message);
        }

        [Fact]
        public void ShouldFailWhenIsLHArrayUsedOnByteArrayParameter()
        {
            MethodInfo method = GetTaskMethodByName(typeof(ParameterTestTasks), "bytes-lh-array-invalid-task");
            ParameterInfo parameter = method.GetParameters()[0];

            var ex = Assert.Throws<LHTaskSchemaMismatchException>(() => new LHTaskParameter(parameter, 0, logger: null));
            Assert.Contains("Cannot use @LHType(isLHArray = true) with byte[]", ex.Message);
        }

        [Fact]
        public void ShouldFailWhenNativeArrayElementResolvesToJsonObj()
        {
            MethodInfo method = GetTaskMethodByName(typeof(ParameterTestTasks), "invalid-native-arr-pojo");
            ParameterInfo parameter = method.GetParameters()[0];

            var ex = Assert.Throws<ArgumentException>(() => new LHTaskParameter(parameter, 0, logger: null));
            Assert.Contains("cannot contain", ex.Message);
        }

        private static MethodInfo GetTaskMethodByName(Type clazz, string taskName)
        {
            return clazz.GetMethods(BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly)
                .Single(m => m.GetCustomAttribute(typeof(LHTaskMethodAttribute)) is LHTaskMethodAttribute attr
                    && attr.Value == taskName);
        }

        private class UnannotatedArrayElement
        {
            public string? Name { get; set; }
        }

        private class ParameterTestTasks
        {
            [LHTaskMethod("test-json-arr")]
            public Task TestJsonArr([LHType(masked: false, name: "param1")] string[] param1)
            {
                return Task.CompletedTask;
            }

            [LHTaskMethod("test-native-arr")]
            public Task TestNativeArr([LHType(masked: false, name: "param1", isLHArray: true)] string[] param1)
            {
                return Task.CompletedTask;
            }

            [LHTaskMethod("test-native-list")]
            public Task TestNativeList([LHType(masked: false, name: "param1", isLHArray: true)] List<string> param1)
            {
                return Task.CompletedTask;
            }

            [LHTaskMethod("non-array-lh-array-invalid-task")]
            public Task NonArrayLhArrayInvalidTask([LHType(masked: false, isLHArray: true)] string customer)
            {
                return Task.CompletedTask;
            }

            [LHTaskMethod("bytes-lh-array-invalid-task")]
            public Task BytesLhArrayInvalidTask([LHType(masked: false, isLHArray: true)] byte[] bytes)
            {
                return Task.CompletedTask;
            }

            [LHTaskMethod("invalid-native-arr-pojo")]
            public Task InvalidNativeArrayPojo([LHType(masked: false, name: "param1", isLHArray: true)] UnannotatedArrayElement[] param1)
            {
                return Task.CompletedTask;
            }
        }
    }
}
